File: spring-rsocket-java-module/src/main/java/com/xunfos/springrsocketjavamodule/controller/TransactionController.java
package com.xunfos.springrsocketjavamodule.controller;

import com.xunfos.springrsocketjavamodule.model.Transaction;
import com.xunfos.springrsocketjavamodule.service.TransactionService;
import lombok.AllArgsConstructor;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.stereotype.Controller;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.time.Duration;
import java.util.List;

@AllArgsConstructor
@Controller
public class TransactionController {
    private TransactionService transactionService;

    @MessageMapping("last")
    public Mono<Transaction> lastTransaction() {
        return Mono.just(transactionService.lastTransaction());
    }

    @MessageMapping("current")
    public Flux<Transaction> current() {
        final List<Transaction> transactionList = transactionService.getCurrentTransactions();

        return Flux.fromStream(transactionList.stream());
    }
}

File: src/main/kotlin/com/xunfos/springrsocketcoroutinesplayground/controller/TransactionController.kt
package com.xunfos.springrsocketcoroutinesplayground.controller

import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.stereotype.Controller
import org.springframework.stereotype.Service


// https://www.baeldung.com/rsocket continue

enum class TransactionType {
    Shopping,
    Education,
    Takeout,
    Transport,
    Other
}

data class Transaction(val type: TransactionType, val amount: Double, val vendor: String)

@Controller
class TransactionController(
    private val transactionService: TransactionService
) {
    @MessageMapping
    fun lastTransaction() = transactionService.lastTransaction()
}

@Service
class TransactionService {
    fun lastTransaction() = Transaction(
        type = TransactionType.Takeout,
        amount = 12.3,
        vendor = "Huzur Kebab"
    )
}
